### ISSUE TITLE
Allow multiple themes to be active at the same time

### ISSUE THREAD

**Issue by bitterulf** [pencilblue/pencilblue#517]
It is really a pain that only the routes of the active theme plugin are used.
If i create two independent plugins wich use different routes for different things i cannot use both at the same time!
Inheriting is not an option because that would add coupling between the independent plugins.

Do i miss a important point here?

Could we have multiple active themes with a ordering to fix that?

**Comment by blakecallens** [comment]
The routes of other plugins work, but the active theme takes priority. For example, if you install the Portfolio theme but don't set it as the active theme, the route of /blog will still work, but it uses the template of 'index'. Since the default theme is active, its index.html template is used instead.

You can work around this by namespacing all your templates to ensure there's no duplication.

**Comment by bitterulf** [comment]
Ok i understand that but let's say you want to alter the rendering of the index in one plugin and the rendering of pages in another plugin. How do you do that? Both needs overwriting core routes.

The reason why i was stumbling across this was that i wanted to create a hooking system
where other plugins can register functions to alter rendering output and such.
These hooks needs to be executed in core controller by overwriting them.

For this to work my hooking plugin needs to be the active theme.
And this means that i can use this general hooking system OR a plugin wich is a theme and not both.

**Comment by blakecallens** [comment]
@brianhyder, would it be possible to write a plugin for the admin section that would allow you to force certain routes to individual plugins?

**Comment by brianhyder** [comment]
No, it isn't but I'm still having a hard time understanding what the issue
is.  In the example of index above you could have:

Plugin A => overrides index controller or templates
Plugin B => overrides controllers or templates for pages

If you set Plugin A to be the active theme then index is taken care of.  As
long as Plugin A does not contain the templates for the pages the request
handler will iterate over the available plugins to find the first one that
contains the route.  Even if the PencilBlue controller is selected the
template service will do the same thing.  It will first look to see if
Plugin A contains the template.  If it does not it will iterate over the
plugins to looking to see who has the template.

I feel like I'm not explaining that well.



On Tue, Mar 31, 2015 at 9:19 AM, Blake Callens <notifications@github.com>
wrote:

> @brianhyder <https://github.com/brianhyder>, would it be possible to
> write a plugin for the admin section that would allow you to force certain
> routes to individual plugins?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pencilblue/pencilblue/issues/517#issuecomment-88081199>
> .
>



-- 
Brian P. Hyder

**Comment by bitterulf** [comment]
I just done some testing and i think it doesn't work as expected:

This is my controller for my first plugin.
```
var CoreIndex = require('../../pencilblue/controllers/index.js');

function HookedIndex(){}

util.inherits(HookedIndex, CoreIndex);

HookedIndex.prototype.render = function(cb) {
  console.log('hook plugin rendering index');
  CoreIndex.prototype.render.apply(this, arguments);
};

HookedIndex.getRoutes = function(cb) {  
	var routes = [
        {
            method: 'get',
            path: "/",
            access_level: 0,
            auth_required: false,
            content_type: 'text/html'
        }
	];
	cb(null, routes);
};

module.exports = HookedIndex;

```
This is only called if the plugin is active AND set as active theme.
If i create a second plugin in the same way just to handle the page rendering only one of them 
(The one wich is set to active theme) works.

```
var CorePage = require('../../pencilblue/controllers/page.js');

function RobIndex(){}

util.inherits(RobIndex, CorePage);

RobIndex.prototype.render = function(cb) {
  console.log('page plugin');
  CorePage.prototype.render.apply(this, arguments);
};

RobIndex.getRoutes = function(cb) {  
	var routes = [
        {
            method: 'get',
            path: "/page/:customUrl",
            access_level: 0,
            auth_required: false,
            content_type: 'text/html'
        }
	];
	cb(null, routes);
};

module.exports = RobIndex;

```
Is that a bug, a feature or a misunderstanding?

And by the way thank you for your time and work you put into the project!

**Comment by blakecallens** [comment]
@brianhyder perhaps the *pencilblue* plugin is taking priority here (being seen as a standard plugin and not the backstop) when the routing is looking to build */page/:customUrl*. If that's the case, is there a way to always push the *pencilblue* plugin to the bottom of the priority list?

**Comment by rainabba** [comment]
I'm playing with an idea and this question leads me to think that in theory, I can assign different themes to different routes and that's what the OP is trying to do, but having trouble with. Correct? Should it be possible? If so, what is the proper mechanism?

I'm looking for a a fairly clean way to dynamically switch the theme based on the hostname (ideally, route as a fallback) used to access the site so that I might host a single site, with multiple subdomains and style each differently?

In other words, I could host with mysite.com and foo.mysite.com and then send someone to foo.mysite.com or foo.mysite.com/foo and give the impression that it's a different site, but without needs to litterally host a 2nd instance of PencilBlue (allowing for content and user sharing)?

**Comment by brianhyder** [comment]
Spoiler alert: multi-tenancy will be in beta in a few weeks :)

**Comment by rainabba** [comment]
@brianhyder I've got an email with Blake where we briefly touched base Aug 6th of last year. Hard to believe it's been that long, but things have progressed and I'm pretty much at the point of no return on a project where I'd like to use PencilBlue and this would be a key feature. I have a couple other questions before diving in and I'm wondering if you've got a paid-support option (or if we could get something together in a hurry) or any other way to connect. If all goes as planned, we would provide an awesome case-study, but I'm on an impossible time-line now and could use any/all help I can get in making my plans a reality. If you have any interest, feel free to drop me a line with my username at gmail. In case a reference is needed, the email I last got from blake was on Mon, Apr 6, 2015 at 6:37 AM.

Sorry if this is inappropriate. I'm actually still learning my way around GitHub in many respects and don't know a better way to connect. I did just reach out to you on GitHub.
